pkg/pillar/kubeapi/storageutils.go

```diff
@@ -0,0 +1,146 @@
+	"golang.org/x/sys/unix"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	longhornDevPath = "/dev/longhorn"
+)
+
+func getMajorMinorStr(stat syscall.Stat_t) string {
+	major := unix.Major(stat.Rdev)
+	minor := unix.Minor(stat.Rdev)
+	return fmt.Sprintf("%d:%d", major, minor)
+}
+
+func isPvAttachedOnThisNode(pv string) bool {
+	if _, err := os.Stat(longhornDevPath + "/" + pv); err != nil {
+		return false
+	}
+	return true
+}
+func isBlockDeviceAttachedOnThisNode(dev string) bool {
+	if _, err := os.Stat("/dev/" + dev); err != nil {
+		return false
+	}
+	return true
+}
+
+// CleanupUnmountedDiskMetrics loops over existing DiskMetric
+// objects and unpublishes them if the device no longer exists
+func CleanupUnmountedDiskMetrics(pubDiskMetric pubsub.Publication, pvcToPvMap map[string]string) {
+	existingMetrics := pubDiskMetric.GetAll()
+
+	for id := range existingMetrics {
+		if strings.HasPrefix(id, "pvc-") {
+			pvName, ok := pvcToPvMap[id]
+			if ok {
+				if !isPvAttachedOnThisNode(pvName) {
+					pubDiskMetric.Unpublish(id)
+				}
+			}
+		} else {
+			if !isBlockDeviceAttachedOnThisNode(id) {
+				pubDiskMetric.Unpublish(id)
+			}
+		}
+	}
+}
+
+// LonghornGetMajorMinorMaps builds two maps between
+// device major:minor -> kube-pv-name/lh-volume-name
+// and kube-pv-name/lh-volume-name -> maj:min to
+// help callers find a PV/PVC in /proc/diskstats
+// which only shows the sdX path.
+func LonghornGetMajorMinorMaps() (map[string]string, map[string]string, error) {
+	lhMajMinToNameMap := make(map[string]string) // maj:min -> kube-pv-name/lh-volume-name
+	lhNameToMajMinMap := make(map[string]string) // kube-pv-name/lh-volume-name -> maj:min
+
+	if _, err := os.Stat(longhornDevPath); err != nil {
+		return lhMajMinToNameMap, lhNameToMajMinMap, fmt.Errorf("longhorn dev path missing")
+	}
+
+	lhPvcList, err := os.ReadDir(longhornDevPath)
+	if err != nil {
+		return lhMajMinToNameMap, lhNameToMajMinMap, fmt.Errorf("unable to read longhorn devs")
+	}
+
+	for _, lhDirEnt := range lhPvcList {
+		var lhStat syscall.Stat_t
+		err := syscall.Stat(longhornDevPath+"/"+lhDirEnt.Name(), &lhStat)
+
+		if err != nil {
+			continue
+		}
+		majMinKey := getMajorMinorStr(lhStat)
+		lhMajMinToNameMap[majMinKey] = lhDirEnt.Name()
+		lhNameToMajMinMap[lhDirEnt.Name()] = majMinKey
+
+	}
+	return lhMajMinToNameMap, lhNameToMajMinMap, nil
+}
+
+// SCSIGetMajMinMaps builds two maps to assist linking with other devices
+// First map: maj:min -> sdX
+// Second map: sdX -> maj:min
+func SCSIGetMajMinMaps() (map[string]string, map[string]string, error) {
+	sdMajMinToNameMap := make(map[string]string) // maj:min -> sdX
+	sdNameToMajMinMap := make(map[string]string) // sdX -> maj:min
+
+	blockDevs, err := os.ReadDir("/sys/class/block/")
+	if err != nil {
+		return sdMajMinToNameMap, sdNameToMajMinMap, fmt.Errorf("unable to read sd devs")
+	}
+
+	for _, devEnt := range blockDevs {
+		if !strings.HasPrefix(devEnt.Name(), "sd") {
```

Comment:
@andrewd-zededa , why are you considering only sdX devices? What about devices that have NVMes that will be named into something like nvme0n1, nvme0n1p1, nvme0n1p2, nvme0n1p3?

Reply:
And also eMMCs, that in this case will be named as something like mmcblk0p1, mmcblk0p2, etc...

Reply:
@rene this is filtering to only sdX devices as those are the only type which are presented as virtual frontend block devices by the longhorn engine we are using currently.  I see your point though and this function could be opened up to build a map of all block devices or allow a filter passed in as a parameter to control the size of the map.